newrelic/hooks/serverless_azure.py

```diff
@@ -0,0 +1,298 @@
+# limitations under the License.
+
+import os
+import re
+import urllib.parse as urlparse
+
+from newrelic.api.application import application_instance
+from newrelic.api.transaction import current_transaction
+from newrelic.api.web_transaction import WebTransaction
+from newrelic.common.object_wrapper import wrap_function_wrapper
+
+
+# TODO: This should serve as a way to determine the trigger type.
+# Right now, we only support HTTP, so this function serves to activate
+# the application if not already registered with the collector as well
+# as determining if this invocation was a cold start or not.
+async def wrap_dispatcher__handle__invocation_request(wrapped, instance, args, kwargs):
+    def bind_params(request, *args, **kwargs):
+        return request
+
+    # Force default registration of the application instance
+    # instead of lazy registration upon the first request
+    app_name = os.environ.get("NEW_RELIC_APP_NAME", os.environ.get("WEBSITE_SITE_NAME", None))
+    application = application_instance(app_name, activate=False)
+    if application and not application.active:
+        application.activate()
+    elif not application:
+        application = application_instance(app_name)
+        application.activate()
+
+    # Logic to determine if this is a cold start since we are not
+    # able to access the logic in the __init__ method of the Dispatcher
+    # class with Python (in the Portal, this is not done in Python)
+    if not hasattr(instance, "_nr_running_dispatcher"):
+        instance._nr_running_dispatcher = True
+        instance._nr_cold_start = True
+
+    request = bind_params(*args, **kwargs)
+
+    if not request:
+        return await wrapped(*args, **kwargs)
+
+    # For now, NR only supports HTTP triggers
+    function_id = request.invocation_request.function_id
+
+    binding_type = instance._functions.get_function(function_id).trigger_metadata["type"]
+    if not binding_type.startswith("http"):
+        return await wrapped(*args, **kwargs)
+
+    return await wrapped(*args, **kwargs)
+
+
+async def wrap_dispatcher__run_async_func(wrapped, instance, args, kwargs):
+    from azure.functions.http import HttpRequest
+
+    def bind_params(context, func, args, *_args, **_kwargs):
+        return context, func, args
+
+    context, func, params = bind_params(*args, **kwargs)
+
+    application = application_instance(
+        os.environ.get("NEW_RELIC_APP_NAME", os.environ.get("WEBSITE_SITE_NAME", None)), activate=False
+    )
+    if application and not application.active:
+        application.activate()
+    elif not application:
+        application = application_instance(
+            os.environ.get("NEW_RELIC_APP_NAME", os.environ.get("WEBSITE_SITE_NAME", None))
+        )
+        application.activate()
+
+    http_request = None
+    for key, value in params.items():
+        if isinstance(value, HttpRequest):
+            http_request = value
+            url_split = urlparse.urlsplit(http_request.url)
```

Comment:
Maybe we could use `getattr` calls here to grab attributes instead of direct access to avoid attribute errors (by potentially also specifying a default value)